data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java

```diff
@@ -162,104 +165,238 @@ public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
+        if (currentBufferPerBatch.getEventCount() > 0) {
+        LOG.info("Force Flushing the remaining {} events in the buffer", currentBufferPerBatch.getEventCount());
+            try {
+                flushToLambdaIfNeeded(resultRecords, true); // Force flush remaining events
+                currentBufferPerBatch.reset();
+            } catch (Exception e) {
+                LOG.error("Exception while flushing remaining events", e);
+                handleFailure(e, currentBufferPerBatch, resultRecords);
+            }
+        }
 
-    }
+        lambdaCommonHandler.waitForFutures(futureList);
 
-    @Override
-    public boolean isReadyForShutdown() {
-        return false;
+        return resultRecords;
     }
 
-    @Override
-    public void shutdown() {
 
-    }
+    void flushToLambdaIfNeeded(List<Record<Event>> resultRecords, boolean forceFlush) {
 
-    protected void flushToLambdaIfNeeded(List<Record<Event>> resultRecords) throws InterruptedException, IOException {
-        LOG.info("Flush to Lambda check: currentBuffer.size={}, currentBuffer.events={}, currentBuffer.duration={}", currentBuffer.getSize(), currentBuffer.getEventCount(), currentBuffer.getDuration());
-
-        if (ThresholdCheck.checkThresholdExceed(currentBuffer, maxEvents, maxBytes, maxCollectionDuration, isBatchEnabled)) {
-            try{
-                codec.complete(currentBuffer.getOutputStream());
-                responsePayload.set(currentBuffer.getPayloadResponseSyncSize());
-                InvokeResponse lambdaResponse = currentBuffer.flushToLambdaSync();
-                handleLambdaResponse(lambdaResponse);
-                numberOfRecordsSuccessCounter.increment(currentBuffer.getEventCount());
-                lambdaLatencyMetric.record(currentBuffer.getFlushLambdaSyncLatencyMetric());
-                requestPayloadMetric.set(currentBuffer.getPayloadRequestSyncSize());
-                Event lambdaEvent = convertLambdaResponseToEvent(lambdaResponse);
-                resultRecords.add(new Record<>(lambdaEvent));
-            } catch(AwsServiceException | SdkClientException e) {
-                LOG.error(EVENT, "Exception occurred while uploading records to lambda. functionName: {} | exception:", functionName, e);
-                numberOfRecordsFailedCounter.increment(currentBuffer.getEventCount());
-            } catch (final IOException e) {
-                LOG.error("Exception while completing codec", e);
-                numberOfRecordsFailedCounter.increment(currentBuffer.getEventCount());
-            }
-            //Reset Buffer
+        LOG.debug("currentBufferPerBatchEventCount:{}, maxEvents:{}, maxBytes:{}, maxCollectionDuration:{}, forceFlush:{} ", currentBufferPerBatch.getEventCount(),maxEvents,maxBytes,maxCollectionDuration, forceFlush);
+        if (forceFlush || ThresholdCheck.checkThresholdExceed(currentBufferPerBatch, maxEvents, maxBytes, maxCollectionDuration)) {
```

Comment:
We should be able to handle this in common with the sink with some slight refactoring.

Reply:
yes it flush method looks very similar and i tried doing that in the earlier PR, but we need to handle release of events differently. For that i introduced isSink flag. But Comment from Krishna was that common code should not have flag like that and i think agree with his point.

Reply:
I agree that a flag is not ideal. But, you can have an injectable strategy. We can resolve this later though.